tensordict/tensordict.py

```diff
@@ -944,18 +944,19 @@ def expand(self, *shape) -> TensorDictBase:
                         new_shape=shape, old_shape=self.batch_size
                     )
                 )
-        for key, value in self.items():
+
+        def _expand_each(value):
             tensor_dims = len(value.shape)
             last_n_dims = tensor_dims - tensordict_dims
             if last_n_dims > 0:
-                d[key] = value.expand(*shape, *value.shape[-last_n_dims:])
+                return value.expand(*shape, *value.shape[-last_n_dims:])
             else:
-                d[key] = value.expand(*shape)
-        return TensorDict(
-            source=d,
-            batch_size=[*shape],
-            device=self.device,
-            _run_checks=False,
+                return value.expand(*shape)
+
+        return _apply_safe(
+            fn=lambda _, value: _expand_each(value),
+            tensordict=self,
+            compute_batch_size=lambda _: shape,
```

Comment:
This will not always give the right answer in the non-auto-nested case. E.g. if we have

```python
td = TensorDict({"a": torch.zeros(3), "b": TensorDict({"c": torch.zeros(3, 4)}, [3, 4])}, [3])
```

Then `td.expand(10, 3)` results in

```
TensorDict(
    fields={
        a: Tensor(shape=torch.Size([10, 3]), device=cpu, dtype=torch.float32, is_shared=False),
        b: TensorDict(
            fields={
                c: Tensor(shape=torch.Size([10, 3, 4]), device=cpu, dtype=torch.float32, is_shared=False)},
            batch_size=torch.Size([10, 3]),
            device=None,
            is_shared=False)},
    batch_size=torch.Size([10, 3]),
    device=None,
    is_shared=False)
```

Note the collapsed batch size of the nested tensordict. I think we're going to need a function that computes the expanded batch size from the nested tensordict, using the `shape`.